docker_image OSX patch

michaeljs1990 opened the pull request:
Check commit enables using tls when using the docker_image module. It
also removes the default for docker_url which doesn't allow us to check
for DOCKER_HOST which is a more sane default. This allows you to use
docker_image on OSX but more documentation is needed.

michaeljs1990 committed:
docker_image TLS

Check commit enables using tls when using the docker_image module. It
also removes the default for docker_url which doesn't allow us to check
for DOCKER_HOST which is a more sane default. This allows you to use
docker_image on OSX but more documentation is needed.

michaeljs1990 commented:
This really needs some documentation because brew as it stands does not make this easy or even possible. I will look into opening a pull request with whoever maintains that repo to see if we can narrow down the issue. Just off the top of my head here is a list of things I had to do to allow this to run after adding TLS.
- Download Python 2.7.10 and install it from the python website
  Before this I got this error on connect even after trying to completely reinstall docker2boot to ensure it wasn't a certificate issue openssl 0.9.8 was installed at this time

``` python
msg: SSLError(SSLError(SSLError(1, u'[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)'),),)
```

upgrading to openssl 1.0.2 the brew default gave the same thing until i tried `pip install requests[security]` which was recommended in a homebrew/python issue. Then the following happened.

``` python
Traceback (most recent call last):
  File "/Users/mschuett/.ansible/tmp/ansible-tmp-1435643901.97-208061863114789/docker_image", line 1937, in <module>
    main()
  File "/Users/mschuett/.ansible/tmp/ansible-tmp-1435643901.97-208061863114789/docker_image", line 303, in main
    images = manager.get_images()
  File "/Users/mschuett/.ansible/tmp/ansible-tmp-1435643901.97-208061863114789/docker_image", line 257, in get_images
    images = self.client.images()
  File "/usr/local/lib/python2.7/site-packages/docker/client.py", line 639, in images
    res = self._result(self._get(self._url("/images/json"), params=params),
  File "/usr/local/lib/python2.7/site-packages/docker/client.py", line 105, in _get
    return self.get(url, **self._set_request_timeout(kwargs))
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 477, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 465, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/usr/local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 544, in urlopen
    body=body, headers=headers)
  File "/usr/local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 341, in _make_request
    self._validate_conn(conn)
  File "/usr/local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 761, in _validate_conn
    conn.connect()
  File "/usr/local/lib/python2.7/site-packages/requests/packages/urllib3/connection.py", line 238, in connect
    ssl_version=resolved_ssl_version)
  File "/usr/local/lib/python2.7/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 255, in ssl_wrap_socket
    ctx = OpenSSL.SSL.Context(_openssl_versions[ssl_version])
KeyError: 5
```
- After Install Python 2.7.10 from python.org I got

```
msg: SSLError(SSLError(CertificateError("hostname '192.168.59.104' doesn't match 'boot2docker'",),),)
```

I added the following to /etc/hosts to fix the issue

```
192.168.59.104  boot2docker
```
- symlink ~/.bash_profile ~/.bashrc since .bash_profile will not be run when ansible executes `ssh localhost /bin/sh`

michaeljs1990 commented:
https://github.com/Homebrew/homebrew-python/issues/232 opened an issue. Unfortunately I don't have the first clue what could cause it so hopefully the package maintainer will have an idea where to start at.

I believe the other error can be fixed in two ways. `-o StrictHostKeyChecking=no` and alerting the user if it fails this could be the issue if OS is Darwin or sending in a patch to boot2docker that would give the user different instructions after starting the VM as it's a manual process right now anyway one extra step wouldn't be egregious.

gregdek commented:
Thanks @michaeljs1990 -- need a bit of guidance here. Do you consider this PR complete enough to be considered for inclusion, or is it blocking for the homebrew patch to be merged upstream? Just trying to figure out how to categorize this PR so we don't lose track of it.

michaeljs1990 commented:
@gregdek I don't personally want to block this issue while waiting on homebrew to possibly fix the issue. What I intend to do is check for failure on a Darwin system and if so display to the user that the python version of brew has some issues and installing from python.org will fix the problem. Once the problem is resolved upstream I will submit another pull request with a documentation fix that alerts the user to update python via brew. What do you think? If you think that is a good course of action I will submit a pull request today with the fix.

gregdek commented:
Sounds like a decent workaround to me.

gregdek commented:
You should just be able to push and this PR should be updated though -- no need for a separate PR.

michaeljs1990 commented:
@gregdek awesome I'll update this tonight. I am doing a little more work with homebrew to ensure i get the documentation right. I'll also run the entire python, ansible install process on a mac that I have available that hasn't had any modifications to it to ensure it's right.

michaeljs1990 committed:
Improve Error Reporting

This will hopefully help mac users be able to quickly resolve any issues
they may find when trying to use this module.

michaeljs1990 committed:
Improve Message

Give user a course of action in the case where the suggestions do not
work. This will hopefully allow us to work through any further issues
much faster.

michaeljs1990 commented:
I think these error messages will prove useful. I am not sure how you feel about having my username in the message but I would love to work through the errors with whoever encounters them to make sure all users can have a pleasant experience and resolve them quickly.

michaeljs1990 committed:
Documentation Fix

Updated documentation to match current module state.

michaeljs1990 committed:
Remove faulty logic

Update logic after splitting the error into two separate messages.

michaeljs1990 commented:
@gregdek this should be good to go now :100:

gregdek commented:
Thx @michaeljs1990.

@ansible/docker folks -- Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please give it a +1; at that point, we will add the “priority_reviewed” flag to this PR, which will move it to the top of our queue for potential inclusion.

michaeljs1990 committed:
remove .geturl()

Can't call geturl on a string.

michaeljs1990 committed:
Handle connection error

Try and help when mac hits a connection error.

michaeljs1990 committed:
Fix message

Previous fix did not actual work. This fix does however.

michaeljs1990 commented:
I was thinking about this some more and I think it might make more sense to improve the docs to let users know that mac osx and boot2docker have these issues instead of littering the source code with comments.

I'll let someone review first before I go changing this PR.

joshuaconner commented:
> It also removes the default for docker_url which doesn't allow us to check
> for DOCKER_HOST which is a more sane default.

I'm not active enough in the ansible world to know any more – is "look it up in the environment" a convention that is used and accepted? My initial instinct is that could result in unexpected issues when managing multiple machines.

michaeljs1990 commented:
@joshuaconner as far as I know it is the only way to ensure this works on a mac (boot2docker) without adding a ton of defaults that you would have to add every time you called the docker image module. The other issue is that if you do set the defaults they can easily change the IP between reboots / machine which would make any module using this hard to port into other environments. 

The docker module uses the same logic as well but I am open to other options I just can't think of any that would work well or don't require extra config.

michaeljs1990 commented:
@ansible/docker I know you guys are super busy but just pinging in case it fell of the radar with how many other PRs you have.

bcoca commented:
the @ansible/docker only works for greg, it is his personal alias ... wish it were usable by all in the project

michaeljs1990 commented:
I was wondering when it didn't show up as highlighted. Github makes it a little hard to handle large projects like this it seems.

michaeljs1990 commented:
@bcoca @gregdek how do i get this back on track for merge into core? Sorry to bug you guys.

abadger commented on cloud/docker/docker_image.py:
```diff
@@ -63,8 +63,43 @@
+        only be used if one of the other tls options require it.
+    choices: [ "no", "encrypt", "verify" ]
+    version_added: "1.9"
+  tls_client_cert:
```
The version_added: entries will need to be updated to "2.0"

abadger commented on cloud/docker/docker_image.py:
```diff
@@ -285,6 +405,45 @@ def main():
+                they are set in .bash_profile you will need to symlink
+                it to .bashrc.
+                '''
+                module.exit_json(failed=True, chaged=manager.has_changed(), msg="SSLError: " + str(e) + environment_error)
```
I see several places where changed is typoed as chaged.   Need to change those.

abadger commented:
@michaeljs1990 sorry it's taken so long to get to this.  I reviewed it today and it looks good except for some minor things that should be easy to correct and a few misgivings about the new error messages.  They seem a bit specific and but my impression is that practically no one is attempting to do docker on Darwin except via boot2docker so perhaps the specificity is acceptable.  They also seem to be giving a lot more recipe-type information than is truly good for an error message and I'm having a harder time justifying that to myself.

Maybe you can move the information into the module documentation's Notes section instead of in the error message?

abadger commented:
Once those things are cleaned up, I think this is fine to merge.

michaeljs1990 commented:
@abadger awesome I'll get this fixed and pushed up tonight. Thanks for the review!

goozbach commented:
@michaeljs1990 any update with this PR? I'd be willing to help if needed.

michaeljs1990 commented:
@goozbach sorry for the delay. Been extremely busy but I hope to finish this up by the end of the week.

abadger commented:
Okay, I made those changes and pushed this into the repository.  My changes are here. https://github.com/ansible/ansible-modules-core/commit/e318be30ff0a4c19e61ff15437ff96a6ba57f696  and your commits show up as the parent of that.  All looks good in devel.  Seems that github doesn't realize that the commits have been merged since I used fastforward to merge it so there's no merge commit.

Thanks @michaeljs1990 for the code and @joshuaconner and @goozbach for reviewing!

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
